core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -0,0 +1,282 @@
+      fields.add(joinField);
+    }
+    Schema keySchema = Schema.recordOf("join.key", fields);
+    StructuredRecord.Builder keyRecordBuilder = StructuredRecord.builder(keySchema);
+    i = 1;
+    for (String joinKey : joinKeys) {
+      keyRecordBuilder.set(String.valueOf(i++), record.get(joinKey));
+    }
+
+    return keyRecordBuilder.build();
+  }
+
+  @Override
+  public JoinConfig getJoinConfig() {
+    return new JoinConfig(requiredInputs);
+  }
+
+  @Override
+  public StructuredRecord merge(StructuredRecord joinKey, Iterable<JoinElement<StructuredRecord>> joinRow) {
+    StructuredRecord.Builder outRecordBuilder = StructuredRecord.builder(outputSchema);
+
+    for (JoinElement<StructuredRecord> joinElement : joinRow) {
+      String stageName = joinElement.getStageName();
+      StructuredRecord record = joinElement.getInputRecord();
+
+      Map<String, String> fieldsToRename = perStageFieldsToRename.row(stageName);
+
+      for (Schema.Field field : record.getSchema().getFields()) {
+        String inputFieldName = field.getName();
+
+        // drop the field if not part of fieldsToRename
+        if (!fieldsToRename.containsKey(inputFieldName)) {
+          continue;
+        }
+
+        String outputFieldName = fieldsToRename.get(inputFieldName);
+        outRecordBuilder.set(outputFieldName, record.get(inputFieldName));
+      }
+    }
+    return outRecordBuilder.build();
+  }
+
+  private void init() {
+    perStageJoinKeys = getJoinKeys();
+    requiredInputs = getRequiredInputs();
+    perStageFieldsToRename = getPerStageFieldsToRename();
+  }
+
+  private Map<String, List<String>> getJoinKeys() {
+    Map<String, List<String>> stageToKey = new HashMap<>();
+    if (!Strings.isNullOrEmpty(conf.getJoinKeys())) {
+      Iterable<String> multipleJoinKeys = Splitter.on('&').trimResults().omitEmptyStrings().split(conf.getJoinKeys());
+      if (Iterables.isEmpty(multipleJoinKeys)) {
+        throw new IllegalArgumentException("Join keys can not be empty.");
+      }
+      int numJoinKeys = 0;
+      for (String singleJoinKey : multipleJoinKeys) {
+        KeyValueListParser kvParser = new KeyValueListParser("\\s*=\\s*", "\\.");
+        Iterable<KeyValue<String, String>> keyValues = kvParser.parse(singleJoinKey);
+        if (numJoinKeys == 0) {
+          numJoinKeys = Iterables.size(keyValues);
+        } else if (numJoinKeys != Iterables.size(keyValues)) {
+          throw new IllegalArgumentException("There should be one join key from each of the stages. Please add join " +
+                                               "keys for each stage.");
+        }
+        for (KeyValue<String, String> keyValue : keyValues) {
+          String stageName = keyValue.getKey();
+          String joinKey = keyValue.getValue();
+          if (!stageToKey.containsKey(stageName)) {
+            stageToKey.put(stageName, new ArrayList<String>());
+          }
+          stageToKey.get(stageName).add(joinKey);
```

Comment:
`KeyValueListParser kvParser = new KeyValueListParser("\\s*=\\s*", "\\.");` strips the prefix.